File: Solution/Programmers/카드 짝 맞추기.py
import itertools
import copy
from collections import deque

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


def distance(start, board):
    q = deque()
    q.append(start)
    dist = [[0] * 4 for _ in range(4)]
    visited = [[False] * 4 for _ in range(4)]
    visited[start[0]][start[1]] = True
    while q:
        y, x = q.popleft()
        for i in range(4):
            ctrl = 1
            while True:
                nx = x + dx[i] * ctrl
                ny = y + dy[i] * ctrl
                if nx < 0 or nx >= 4 or ny < 0 or ny >= 4:
                    ctrl -= 1
                    break
                if board[ny][nx]:
                    break
                ctrl += 1
            for j in [1, ctrl]:
                nx = x + dx[i] * j
                ny = y + dy[i] * j
                if nx < 0 or nx >= 4 or ny < 0 or ny >= 4 or visited[ny][nx]:
                    continue
                dist[ny][nx] = dist[y][x] + 1
                visited[ny][nx] = True
                q.append([ny, nx])

    return dist


def solution(board, r, c):
    card_cnt = 0
    card_info = {}
    for i in range(4):
        for j in range(4):
            if board[i][j]:
                card_cnt = max(card_cnt, board[i][j])
                if card_info.get(board[i][j]) is not None:
                    card_info[board[i][j]].append((i, j))
                else:
                    card_info[board[i][j]] = [(i, j)]

    card_case = list(itertools.permutations(range(1, card_cnt + 1), card_cnt))
    ans = int(1e9)
    for case in card_case:
        curr_pos = [r, c]
        copy_board = copy.deepcopy(board)
        result = 0
        for card in case:
            tmp = [0, 0]
            for i in range(2):
                dist = distance(curr_pos, copy_board)
                tmp[i] = dist[card_info[card][i][0]][card_info[card][i][1]]
            for i in range(2):
                curr_pos = card_info[card][i]
                dist = distance(curr_pos, copy_board)
                tmp[i] += dist[card_info[card][(i + 1) % 2][0]][
                    card_info[card][(i + 1) % 2][1]
                ]
            if tmp[0] < tmp[1]:
                curr_pos = card_info[card][1]
                result += tmp[0] + 2
            else:
                curr_pos = card_info[card][0]
                result += tmp[1] + 2
            a, b = card_info[card][0], card_info[card][1]
            copy_board[a[0]][a[1]] = 0
            copy_board[b[0]][b[1]] = 0
        ans = min(ans, result)

    return ans

File: Solution/Programmers/카드 짝 맞추기.js
const dx = [1, -1, 0, 0];
const dy = [0, 0, 1, -1];

function getPermutations(arr, selectNumber) {
  const results = [];
  if (selectNumber === 1) return arr.map(value => [value]);

  arr.forEach((fixed, index, origin) => {
    const rest = [...origin.slice(0, index), ...origin.slice(index + 1)];
    const permutations = getPermutations(rest, selectNumber - 1);
    const attached = permutations.map(permutation => [fixed, ...permutation]);
    results.push(...attached);
  });

  return results;
}

function bfs(board, start) {
  let q = [];
  let visited = Array.from(Array(4), () => Array(4).fill(false));
  let dist = Array.from(Array(4), () => Array(4).fill(0));
  visited[start[0]][start[1]] = true;
  q.push(start);
  while (q.length) {
    const curr = q.shift();
    for (let i = 0; i < 4; i++) {
      let ctrl = 1;
      while (true) {
        const nx = curr[1] + dx[i] * ctrl;
        const ny = curr[0] + dy[i] * ctrl;

        if (nx < 0 || nx >= 4 || ny < 0 || ny >= 4) {
          ctrl -= 1;
          break;
        }
        if (board[ny][nx]) break;

        ctrl += 1;
      }

      for (const j of [1, ctrl]) {
        const nx = curr[1] + dx[i] * j;
        const ny = curr[0] + dy[i] * j;

        if (nx < 0 || nx >= 4 || ny < 0 || ny >= 4 || visited[ny][nx]) continue;
        dist[ny][nx] = dist[curr[0]][curr[1]] + 1;
        visited[ny][nx] = true;
        q.push([ny, nx]);
      }
    }
  }

  return dist;
}

function solution(board, r, c) {
  const cardList = {};
  let cardNumbers = [];
  for (let i = 0; i < 4; i++) {
    for (let j = 0; j < 4; j++) {
      if (board[i][j] === 0) continue;
      if (board[i][j] in cardList) {
        cardList[board[i][j]].push([i, j]);
      } else {
        cardList[board[i][j]] = [[i, j]];
      }

      cardNumbers.push(board[i][j]);
    }
  }
  cardNumbers = [...new Set(cardNumbers)].sort();

  let ans = Number.POSITIVE_INFINITY;
  const cardSequence = getPermutations(cardNumbers, cardNumbers.length);
  for (const seq of cardSequence) {
    let result = 0;
    let currPos = [r, c];
    const boardCopy = JSON.parse(JSON.stringify(board));
    for (const targetCard of seq) {
      let tmp = [0, 0];
      for (let i = 0; i < 2; i++) {
        let dist = bfs(boardCopy, currPos);
        tmp[i] = dist[cardList[targetCard][i][0]][cardList[targetCard][i][1]];
      }
      for (let i = 0; i < 2; i++) {
        currPos = cardList[targetCard][i];
        let dist = bfs(boardCopy, currPos);
        tmp[i] +=
          dist[cardList[targetCard][(i + 1) % 2][0]][
            cardList[targetCard][(i + 1) % 2][1]
          ];
      }
      if (tmp[0] < tmp[1]) {
        currPos = cardList[targetCard][1];
        result += tmp[0] + 2;
      } else {
        currPos = cardList[targetCard][0];
        result += tmp[1] + 2;
      }
      const card1 = cardList[targetCard][0];
      const card2 = cardList[targetCard][1];

      boardCopy[card1[0]][card1[1]] = 0;
      boardCopy[card2[0]][card2[1]] = 0;
    }
    ans = Math.min(ans, result);
  }

  return ans;
}
